drv/psc-psu-update/src/main.rs

```diff
@@ -0,0 +1,262 @@
+    devices::mwocp68_psu2mcu,
+    devices::mwocp68_psu3mcu,
+    devices::mwocp68_psu4mcu,
+    devices::mwocp68_psu5mcu,
+];
+
+static PSU: ClaimOnceCell<[Psu; 6]> = ClaimOnceCell::new(
+    [Psu {
+        last_checked: None,
+        present: None,
+        power_good: None,
+        firmware_matches: None,
+        firmware_revision: None,
+        update_started: None,
+        update_succeeded: None,
+        update_failure: None,
+        update_backoff: None,
+    }; 6],
+);
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+enum Trace {
+    None,
+    Start,
+    PowerGoodFailed(u8, drv_i2c_devices::mwocp68::Error),
+    FirmwareRevFailed(u8, drv_i2c_devices::mwocp68::Error),
+    Psu(u8),
+    AttemptingUpdate(u8),
+    BackingOff(u8),
+    UpdateFailed,
+    UpdateFailedState(Option<UpdateState>),
+    UpdateFailure(Mwocp68Error),
+    UpdateState(UpdateState),
+    WroteBlock,
+    UpdateSucceeded,
+    UpdateDelay(u64),
+}
+
+const MWOCP68_FIRMWARE_REV: Mwocp68FirmwareRev = Mwocp68FirmwareRev(*b"0762");
+const MWOCP68_FIRMWARE_PAYLOAD: &'static [u8] =
+    include_bytes!("mwocp68-0762.bin");
+
+ringbuf!(Trace, 32, Trace::None);
+
+#[derive(Copy, Clone, PartialOrd, PartialEq)]
+struct Ticks(u64);
+
+impl Ticks {
+    fn now() -> Self {
+        Self(sys_get_timer().now)
+    }
+}
+
+impl Add for Ticks {
+    type Output = Self;
+
+    fn add(self, other: Self) -> Self {
+        Self(self.0 + other.0)
+    }
+}
+
+#[derive(Copy, Clone, Default)]
+struct Psu {
+    /// When did we last check this device?
+    last_checked: Option<Ticks>,
+
+    /// Is the device physically present?
+    present: Option<bool>,
+
+    /// Is the device on and with POWER_GOOD set?
+    power_good: Option<bool>,
+
+    /// The last firmware revision read
+    firmware_revision: Option<Mwocp68FirmwareRev>,
+
+    /// Does the firmware we have match the firmware here?
+    firmware_matches: Option<bool>,
+
+    /// What time did we start an update?
+    update_started: Option<Ticks>,
+
+    /// What time did the update complete?
+    update_succeeded: Option<Ticks>,
+
+    /// What time did the update last fail, if any?
+    update_failure: Option<(Ticks, Option<UpdateState>, Mwocp68Error)>,
+
+    /// How long should the next update backoff, if at all? (In ticks.)
+    update_backoff: Option<Ticks>,
+}
+
+impl Psu {
+    fn update_should_be_attempted(&mut self, dev: &Mwocp68, ndx: u8) -> bool {
+        let now = Ticks::now();
+
+        self.last_checked = Some(now);
+        self.power_good = None;
+        self.firmware_matches = None;
+        self.firmware_revision = None;
+
+        if !dev.present() {
+            self.present = Some(false);
+
+            //
+            // If we are seeing our device as not present, we will clear our
+            // backoff value: if/when a PSU is plugged back in, we want to
+            // attempt to update it immediately if the firmware revision
+            // doesn't match our payload.
```

Comment:
This doesn't seem totally robust: if someone swaps a PSU during the 10-second wait between checks, we won't notice that it's gone (and therefore will use the original `update_backoff`).  Should we be storing the serial number (`MFR_SERIAL`) to catch this edge case?

Reply:
Yes, but also:

![2024-11-22-19-59-46_1041x31](https://github.com/user-attachments/assets/e0e7d6e6-b9fa-4ae0-a645-075d6ff8dea0)

At least they define the length! And hard to see how it will be worse to track it...

Reply:
Agreed -- fixed.